src/CharRNN/index.js

```diff
@@ -0,0 +1,245 @@
+    this.vocabSize = Object.keys(json).length;
+  }
+
+  async initCells() {
+    this.cells = [];
+    this.zeroState = { c: [], h: [] };
+    const forgetBias = tf.tensor(1.0);
+
+    const lstm = (i) => {
+      const cell = (DATA, C, H) =>
+        tf.basicLSTMCell(forgetBias, this.model[`Kernel_${i}`], this.model[`Bias_${i}`], DATA, C, H);
+      return cell;
+    };
+
+    for (let i = 0; i < this.cellsAmount; i += 1) {
+      this.zeroState.c.push(tf.zeros([1, this.model[`Bias_${i}`].shape[0] / 4]));
+      this.zeroState.h.push(tf.zeros([1, this.model[`Bias_${i}`].shape[0] / 4]));
+      this.cells.push(lstm(i));
+    }
+
+    this.state = this.zeroState;
+  }
+
+
+  async generateInternal(options) {
+    await this.ready;
+    const seed = options.seed || this.defaults.seed;
+    const length = +options.length || this.defaults.length;
+    const temperature = +options.temperature || this.defaults.temperature;
+    const stateful = options.stateful || this.defaults.stateful;
+    if (!stateful) {
+      this.state = this.zeroState;
+    }
+
+    const results = [];
+    const userInput = Array.from(seed);
+    const encodedInput = [];
+
+    userInput.forEach((char) => {
+      encodedInput.push(this.vocab[char]);
+    });
+
+    let input = encodedInput[0];
+    let probabilitiesNormalized = []; // will contain final probabilities (normalized)
+
+    for (let i = 0; i < userInput.length + length + -1; i += 1) {
+      const onehotBuffer = tf.buffer([1, this.vocabSize]);
+      onehotBuffer.set(1.0, 0, input);
+      const onehot = onehotBuffer.toTensor();
+      let output;
+      if (this.model.embedding) {
+        const embedded = tf.matMul(onehot, this.model.embedding);
+        output = tf.multiRNNCell(this.cells, embedded, this.state.c, this.state.h);
+      } else {
+        output = tf.multiRNNCell(this.cells, onehot, this.state.c, this.state.h);
+      }
+
+      this.state.c = output[0];
+      this.state.h = output[1];
+
+      const outputH = this.state.h[1];
+      const weightedResult = tf.matMul(outputH, this.model.fullyConnectedWeights);
+      const logits = tf.add(weightedResult, this.model.fullyConnectedBiases);
+      const divided = tf.div(logits, tf.tensor(temperature));
+      const probabilities = tf.exp(divided);
+      probabilitiesNormalized = await tf.div(
+        probabilities,
+        tf.sum(probabilities),
+      ).data();
+
+      if (i < userInput.length - 1) {
+        input = encodedInput[i + 1];
+      } else {
+        input = sampleFromDistribution(probabilitiesNormalized);
+        results.push(input);
+      }
+    }
+
+    let generated = '';
+    results.forEach((char) => {
+      const mapped = Object.keys(this.vocab).find(key => this.vocab[key] === char);
+      if (mapped) {
+        generated += mapped;
+      }
+    });
+    this.probabilities = probabilitiesNormalized;
+    return {
+      sample: generated,
+      state: this.state,
+      probabilities: this.probabilities,
```

Comment:
The probabilities as is aren't useful to the end user. I think if we are planning to return them we'll want to reformat them into a nice object with an array of characters and probabilities? Now that I think about it, it probably only makes sense to return the probabilities in the "stateful" case as what are the probabilities when you are generating full text?

Reply:
fixed